File: backend/app/schemas/market.py
"""
Market Data Schemas: Mandi prices, MSP benchmarks, and price analytics.
"""

import datetime
from typing import Optional

from pydantic import BaseModel, Field


# ── Mandi Price Models ────────────────────────────────────────────────


class MandiPrice(BaseModel):
    """A single commodity price record from a mandi."""

    state: str
    district: str
    market: Optional[str] = None
    commodity: str
    commodity_normalized: Optional[str] = None
    variety: Optional[str] = None
    grade: Optional[str] = None
    arrival_date: datetime.date
    min_price: Optional[float] = None
    max_price: Optional[float] = None
    modal_price: float


class MandiPriceResponse(BaseModel):
    """Response for mandi price queries."""

    state: Optional[str] = None
    district: Optional[str] = None
    total: int
    prices: list[MandiPrice] = Field(default_factory=list)
    source: str = "data.gov.in (Ministry of Agriculture)"


# ── Price Trend Models ────────────────────────────────────────────────


class PriceTrendPoint(BaseModel):
    """A single point in a price trend over time."""

    price_date: datetime.date = Field(alias="date")
    avg_modal_price: float
    min_price: Optional[float] = None
    max_price: Optional[float] = None
    record_count: int = 0


class PriceTrend(BaseModel):
    """Price trend for a commodity in a state."""

    state: str
    commodity: str
    period_start: Optional[datetime.date] = None
    period_end: Optional[datetime.date] = None
    data_points: list[PriceTrendPoint] = Field(default_factory=list)
    avg_price: Optional[float] = None
    price_change_pct: Optional[float] = None


# ── MSP Comparison Models ────────────────────────────────────────────


class MSPRate(BaseModel):
    """Official Minimum Support Price for a crop."""

    crop: str
    season: str
    year: int
    msp_price: float
    grade: Optional[str] = None
    unit: str = "INR/quintal"


class MSPComparisonItem(BaseModel):
    """A district's market price compared to MSP."""

    district: str
    market: Optional[str] = None
    avg_modal_price: float
    msp_price: float
    price_vs_msp_ratio: float = Field(
        ..., description="Ratio: > 1.0 means farmer gets above MSP"
    )
    premium_or_deficit_pct: float = Field(
        ..., description="Percentage above(+) or below(-) MSP"
    )
    status: str = Field(
        ..., description="'Above MSP', 'At MSP', or 'Below MSP'"
    )


class MSPComparisonResponse(BaseModel):
    """MSP comparison results for a state and crop."""

    state: str
    crop: str
    year: int
    msp: MSPRate
    districts: list[MSPComparisonItem] = Field(default_factory=list)
    state_avg_modal_price: Optional[float] = None
    state_avg_ratio: Optional[float] = None
    districts_above_msp: int = 0
    districts_below_msp: int = 0
    source: str = "data.gov.in + CACP"


# ── Price Map Models ─────────────────────────────────────────────────


class PriceMapItem(BaseModel):
    """District-level price for map choropleth."""

    state: str
    district: str
    commodity: str
    avg_modal_price: float
    min_price: Optional[float] = None
    max_price: Optional[float] = None
    record_count: int = 0
    price_date: Optional[datetime.date] = Field(None, alias="date")


class PriceMapResponse(BaseModel):
    """Price map data for a commodity across all districts."""

    commodity: str
    total_districts: int
    items: list[PriceMapItem] = Field(default_factory=list)
    price_range: dict[str, float] = Field(
        default_factory=dict,
        description="min/max/avg across all districts",
    )


# ── Available Commodities ────────────────────────────────────────────


class CommodityInfo(BaseModel):
    """Info about an available commodity in the market data."""

    name: str
    normalized: str
    record_count: int
    states_count: int
    latest_date: Optional[datetime.date] = None
    avg_price: Optional[float] = None

File: frontend/src/app/services/api/market.ts
import { buildPublicApiV1Url } from './config';

export interface CommodityInfo {
    name: string;
    normalized: string;
    record_count: number;
    states_count: number;
    latest_date: string | null;
    avg_price: number | null;
}

export interface PriceTrendPoint {
    date: string;
    avg_modal_price: number;
    min_price: number | null;
    max_price: number | null;
    record_count: number;
}

export interface PriceTrend {
    state: string;
    commodity: string;
    period_start: string | null;
    period_end: string | null;
    data_points: PriceTrendPoint[];
    avg_price: number | null;
    price_change_pct: number | null;
}

export interface MSPRate {
    crop: string;
    season: string;
    year: number;
    msp_price: number;
    grade: string | null;
    unit: string;
}

export interface MSPComparisonItem {
    district: string;
    market: string | null;
    avg_modal_price: number;
    msp_price: number;
    price_vs_msp_ratio: number;
    premium_or_deficit_pct: number;
    status: 'Above MSP' | 'At MSP' | 'Below MSP';
}

export interface MSPComparisonResponse {
    state: string;
    crop: string;
    year: number;
    msp: MSPRate;
    districts: MSPComparisonItem[];
    state_avg_modal_price: number | null;
    state_avg_ratio: number | null;
    districts_above_msp: number;
    districts_below_msp: number;
    source: string;
}


export async function fetchAvailableCommodities(): Promise<CommodityInfo[]> {
    const response = await fetch(buildPublicApiV1Url('/market/commodities'), { cache: 'no-store' });
    if (!response.ok) {
        throw new Error(`Failed to fetch commodities: ${response.statusText}`);
    }
    return response.json();
}

export async function fetchPriceTrends(state: string, commodity: string, days: number = 30): Promise<PriceTrend> {
    const url = new URL(buildPublicApiV1Url('/market/trends'));
    url.searchParams.append('state', state);
    url.searchParams.append('commodity', commodity);
    url.searchParams.append('days', days.toString());

    const response = await fetch(url.toString(), { cache: 'no-store' });
    if (!response.ok) {
        throw new Error(`Failed to fetch price trends: ${response.statusText}`);
    }
    return response.json();
}

export async function fetchMSPComparison(state: string, crop: string, year?: number): Promise<MSPComparisonResponse> {
    const url = new URL(buildPublicApiV1Url('/market/msp-comparison'));
    url.searchParams.append('state', state);
    url.searchParams.append('crop', crop);
    if (year) {
        url.searchParams.append('year', year.toString());
    }

    const response = await fetch(url.toString(), { cache: 'no-store' });
    if (!response.ok) {
        throw new Error(`Failed to fetch MSP comparison: ${response.statusText}`);
    }
    return response.json();
}
